### PR TITLE
Feature/bp 1262 remove duplicates from gateway buckaroo classes

### PR DESCRIPTION
**Pull request by serpentscode**

### CODE REVIEW

File: gateway-buckaroo.php [hunk 1]
```diff
@@ -1192,6 +1192,233 @@ protected function process_refund_common($action, $order_id, $amount = null, $re
...
+                $products[]                = $tmp;
+                $itemsTotalAmount -= 0.01;
+            }
+        }
+
+        return $products;
+    }
+
+    public function getProductImage($product, $method, $sendImageInfo = null) {
+        if ($method != 'afterpay-new' || $method != 'klarna') {
```
Comment (vladislav-padalka-hys) [inline gateway-buckaroo.php:1405]: I don't think it's a good idea to place specific payment method implementation into common class
What if we will specific logic to descendants by redefining /extending methods inside them ?

File: gateway-buckaroo.php [hunk 2]
```diff
@@ -1192,6 +1192,233 @@ protected function process_refund_common($action, $order_id, $amount = null, $re
...
+            } elseif ($afterpay_version == 'afterpay-new' && $method == 'partial_refunds'){
+                $tmp["ArticleVatcategory"] = WC_Tax::_get_tax_rate($shippingTaxClassKey)['tax_rate'] ?? 0;
+            }
+
+            return ['costs' => $shippingCosts, 'shipping_virtual_product' => $tmp];
+        }
+        return ['costs' => 0];
+    }
+
+    public function getProductsInfo($order, $amountDedit, $shippingCosts, $payment_method, $sendImageInfo = null){
```
Comment (vladislav-padalka-hys) [inline gateway-buckaroo.php:1269]: that method became to be quite huge and complex 
could you please split it into several logical chunks
also maybe to move specific payment methods logic into corresponding descendants classes ?

**Review by vladislav-padalka-hys — CHANGES_REQUESTED** [review]

File: gateway-buckaroo.php (hunk 2 above)
Comment (serpentscode) [reply gateway-buckaroo.php:1269]: Yes, I wasn't too happy with the result too, things got a bit chunky and with some specifics. I'll rework this method to split some functionality out, keep most of common code here and move just specific back into each payment method. This also should solve your first input regarding getProductImage..

Thanks!

Comment (vladislav-padalka-hys) [reply gateway-buckaroo.php:1269]: Thank you ! Then result will be really great !

**Review by vladislav-padalka-hys — APPROVED** [review]

**Pull request merged**